peregrine/resources/submission/graphql/node.py

```diff
@@ -788,3 +786,97 @@ def get_fields():
+            schema = dictionary.schema[node]
+            if schema['category'].endswith('_file'):
+                fields = schema['properties'].keys()
+                if shd_fields is None:
+                    shd_fields = set(fields)
+                else:
+                    shd_fields = shd_fields.intersection(fields)
+        shd_fields_dict = {}
+        for field in shd_fields:
+            if field == 'file_size':
+                shd_fields_dict[field] = graphene.Int()
+            # elif field.endswith('_id'):
+            #     shd_fields_dict[field] = graphene.ID()
+            else:
+                shd_fields_dict[field] = graphene.String()
+        DataNode.shared_fields = shd_fields_dict
+    return DataNode.shared_fields
+
+
+def resolve_datanode(self, info, **args):
+    """The root query for the :class:`DataNode` node interface.
+
+    :returns:
+        A list of graphene object classes.
+
+    """
+
+    # get the list of categories that are data categories
+    data_types = filter(lambda node: dictionary.schema[node]['category'].endswith('_file'), dictionary.schema)
+
+    # get the subclasses for the data categories
+    data_types = filter(lambda node: node.label in data_types, psqlgraph.Node.get_subclasses())
+
+    q_all = []
+    for data_type in data_types:
+
+        q = get_authorized_query(data_type)
+        if 'project_id' in args:
+            q = q.filter(q.entity()._props['project_id'].astext
+                         == args['project_id'])
+
+        q = apply_query_args(q, args, info)
+
+        if 'of_type' in args:
+            # TODO: (jsm) find a better solution.  currently this filter
+            # will do a subquery for each type AND LOAD THE IDS of all the
+            # nodes, then perform a second query given those ids.  We
+            # cannot do a ``select_from`` because it does not work
+            # properly for the abstract base class with concrete table
+            # inheritance (a.k.a it can't find the colums for Node)
+            of_types = set(args['of_type'])
+            entities = [psqlgraph.Node.get_subclass(label) for label in of_types]
+            entities = [e for e in entities if e]
+
+            ids = []
+            for label in of_types:
+                entity = psqlgraph.Node.get_subclass(label)
+                q = get_authorized_query(entity)
+                q = apply_query_args(q, args, info)
+                try:
+                    ids += [n.node_id for n in q.all()]
+                except Exception as e:
+                    capp.logger.exception(e)
+                    raise
+            q = get_authorized_query(psqlgraph.Node).ids(ids)
+            q = apply_arg_limit(q, args, info)
+            q = apply_arg_offset(q, args, info)
+
+        q_all.extend(q.all())
+
+    return [__gql_object_classes[n.label](**load_node(n, info)) for n in q_all]
+
+
+def get_datanode_interface_args():
+    to_add = dict(dict(of_type=graphene.List(graphene.String),
```

Comment:
this is kind of hard to read; something like this might be more clear:
```python
args = get_base_node_args()
args.update(get_shared_fields_dict())
args.update({
    'of_type': graphene.List(graphene.String),
    'project_id': graphene.String(),
})
return args
```